### ISSUE TITLE
lambda input type is `Any` when function type is a union

### ISSUE THREAD

**Issue by DetachHead** [python/mypy#13036]
```py
from typing import Callable
def f(a: Callable[[int], None] | Callable[[int, str], None]) -> None: ...
    
f(lambda x: reveal_type(x))  # Any
```
[playground](https://mypy-play.net/?mypy=master&python=3.10&flags=show-error-codes%2Callow-redefinition%2Cstrict%2Ccheck-untyped-defs%2Cdisallow-any-decorated%2Cdisallow-any-expr%2Cdisallow-any-explicit%2Cdisallow-any-generics%2Cdisallow-any-unimported%2Cdisallow-incomplete-defs%2Cdisallow-subclassing-any%2Cdisallow-untyped-calls%2Cdisallow-untyped-decorators%2Cdisallow-untyped-defs%2Cno-implicit-optional%2Cno-implicit-reexport%2Cstrict-equality%2Cwarn-incomplete-stub%2Cwarn-redundant-casts%2Cwarn-return-any%2Cwarn-unreachable%2Cwarn-unused-configs%2Cwarn-unused-ignores&gist=c54846c0b9d860f266c491008ccaeb89)

**Comment by erictraut** [comment]
The title of this bug is misleading. The real bug here is that mypy doesn't report a type violation for `f(lambda x: x)`. This lambda doesn't satisfy either `Callable[[int], None]` or `Callable[[int, str], None]`, so it should be flagged as a type violation.

**Comment by DetachHead** [comment]
it still happens when the lambda does satisfy one of the types in the union:

```py
from typing import Callable
def f(a: Callable[[int], int] | Callable[[int, str], None]) -> None: ...

f(lambda x: reveal_type(x))  # Any
```

i guess that was a mistake in my original post but i guess that means there are two different issues here
